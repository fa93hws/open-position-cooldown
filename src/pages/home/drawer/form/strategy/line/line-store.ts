import { makeObservable, computed } from 'mobx';

import { InputStore } from '@ui/text-input/input-store';

export class LineStore {
  private readonly priceStore: InputStore;

  private readonly quantityStore: InputStore;

  constructor({
    priceStore,
    quantityStore,
  }: {
    priceStore: InputStore;
    quantityStore: InputStore;
  }) {
    this.priceStore = priceStore;
    this.quantityStore = quantityStore;
    makeObservable(this, {
      price: computed,
      quantity: computed,
    });
  }

  get price() {
    return parseFloat(this.priceStore.value);
  }

  get quantity() {
    return parseInt(this.quantityStore.value, 10);
  }
}
