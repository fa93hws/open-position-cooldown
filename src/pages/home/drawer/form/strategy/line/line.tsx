import * as React from 'react';
import { Box, Typography } from '@material-ui/core';
import { makeStyles, withTheme, WithTheme } from '@material-ui/core/styles';
import { observer } from 'mobx-react';

import { createInput, ExposedInputProps } from '@ui/text-input/input';
import { nonEmpty, mustBeNumber } from '@ui/text-input/validator';
import { LineStore } from './line-store';

type LineProps = WithTheme & {
  PriceInput: React.ComponentType<ExposedInputProps>;
  QuantityInput: React.ComponentType<ExposedInputProps>;
};

export const useInputStyles = makeStyles((theme) => ({
  root: {
    height: 28,
  },
  input: {
    padding: theme.spacing(1),
  },
}));

export const Line = React.memo(
  withTheme((props: LineProps) => {
    const inputStyles = useInputStyles(props.theme);
    return (
      <Box display="flex" alignItems="center">
        <Typography variant="body1" component="span">
          跌至
        </Typography>
        <Box width={100} mx={1}>
          <props.PriceInput inputClasses={inputStyles} />
        </Box>
        <Typography variant="body1" component="span">
          买入
        </Typography>
        <Box width={80} mx={1}>
          <props.QuantityInput inputClasses={inputStyles} />
        </Box>
        <Typography variant="body1" component="span">
          股
        </Typography>
      </Box>
    );
  }),
);

export function createLine(): [React.ComponentType, LineStore] {
  const [PriceInput, priceStore] = createInput([nonEmpty, mustBeNumber]);
  const [QuantityInput, quantityStore] = createInput([nonEmpty, mustBeNumber]);
  const store = new LineStore({ priceStore, quantityStore });
  const Component = observer(() => (
    <Line PriceInput={PriceInput} QuantityInput={QuantityInput} />
  ));
  return [Component, store];
}
