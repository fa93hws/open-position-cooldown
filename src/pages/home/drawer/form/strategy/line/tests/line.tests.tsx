import * as React from 'react';
import { shallow } from 'enzyme';

import { createLine, Line } from '../line';

describe('Line', () => {
  it('renders the following', () => {
    const PriceInput = () => <div>PriceInput</div>;
    const QuantityInput = () => <div>QuantityInput</div>;
    expect(
      <Line PriceInput={PriceInput} QuantityInput={QuantityInput} />,
    ).toMatchRenderedSnapshot();
  });

  it('is able to be mounted', () => {
    const [Component] = createLine();
    expect(() => shallow(<Component />)).not.toThrowError();
  });
});
