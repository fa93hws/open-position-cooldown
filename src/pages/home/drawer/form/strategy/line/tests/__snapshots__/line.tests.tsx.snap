// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line renders the following 1`] = `
<div
  class="MuiBox-root MuiBox-root-3"
>
  <span
    class="MuiTypography-root MuiTypography-body1"
  >
    跌至
  </span>
  <div
    class="MuiBox-root MuiBox-root-4"
  >
    <div>
      PriceInput
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-body1"
  >
    买入
  </span>
  <div
    class="MuiBox-root MuiBox-root-5"
  >
    <div>
      QuantityInput
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-body1"
  >
    股
  </span>
</div>
`;
